### ISSUE TITLE
Improve vectorized operations of `GroupColumn`

### ISSUE THREAD

**Issue by Rachelint** [apache/datafusion#13275]
### Is your feature request related to a problem or challenge?

Some good points about improving vectorized operations of `GroupColumn` from @Dandandan and @jayzhan211 in #12996 :

https://github.com/apache/datafusion/pull/12996#discussion_r1830810815
https://github.com/apache/datafusion/pull/12996#discussion_r1818475481
https://github.com/apache/datafusion/pull/12996#discussion_r1819225765
https://github.com/apache/datafusion/pull/12996#discussion_r1823772169

This pr is for tracking such points for conintuing improving performace of multi group by.

**Comment by Rachelint** [comment]
I tried the possible optimizations for`vectorized_equal_to`:
https://github.com/apache/datafusion/pull/12996#discussion_r1818601807
https://github.com/apache/datafusion/pull/12996#discussion_r1823151542

But unfortunately seems no obvious improvement in clickbench, tpch and so on... although some improvement can see in benchmark for the function in https://github.com/apache/datafusion/pull/13539

For `take + eq`, use `eq` in arrow kernel indeed can get performance improve, but the cost of `take` seems unacceptable...

For optimizations for special case of `input array` (like `all rows in the array needed`, `no nulls although it is nullable`), it can see some improvement in the benchmark for function, but see no improvement in e2e benchmark(clickbench, tpch). I guess they may be not the hot path...

I plan to not make further attempts recently...

cc @Dandandan @alamb @jayzhan211

**Comment by Rachelint** [comment]
> I tried the possible optimizations for`vectorized_equal_to`: [#12996 (comment)](https://github.com/apache/datafusion/pull/12996#discussion_r1818601807) [#12996 (comment)](https://github.com/apache/datafusion/pull/12996#discussion_r1823151542)
> 
> But unfortunately seems no obvious improvement in clickbench, tpch and so on... although some improvement can see in benchmark for the function in #13539
> 
> For `take + eq`, use `eq` in arrow kernel indeed can get performance improve, but the cost of `take` seems unacceptable...
> 
> For optimizations for special case of `input array` (like `all rows in the array needed`, `no nulls although it is nullable`), it can see some improvement in the benchmark for function, but see no improvement in e2e benchmark(clickbench, tpch). I guess they may be not the hot path...
> 
> I plan to not make further attempts recently...
> 
> cc @Dandandan @alamb @jayzhan211

@LeslieKid maybe you can try it if possible optimizations not included above.

**Comment by Dandandan** [comment]
Thanks @Rachelint for summarizing, that's interesting. One big difference between `take` + `eq`  in join versus grouped aggregates seems to be that the `ValueBuilder`s already are type erased, so maybe there is nothing to gain.

It seems for the join code, this could actually be used more (and joining on string columns might be not so fast at the moment).

**Comment by Rachelint** [comment]
> Thanks @Rachelint for summarizing, that's interesting. One big difference between `take` + `eq` in join versus grouped aggregates seems to be that the `ValueBuilder`s already are type erased, so maybe there is nothing to gain (and using `take` and `eq` just creates an extra copy.
> 
> It seems for the join code, this could actually be used more (and joining on string columns might be not so fast at the moment).

🤔 I think maybe we can indeed try some solutions without `take` in join? @LeslieKid seems trying it.

**Comment by Dandandan** [comment]
Oh I saw @LeslieKid already commented on that issue 👍

**Comment by LeslieKid** [comment]
> 🤔 I think maybe we can indeed try some solutions without `take` in join? @LeslieKid seems trying it.

Yes, I am working on comparing arrays with indices in a for-loop (similar to the implementation of `vectorized_equal_to()` in multi group-by) instead of using `take` in hash join. See #13607
